base/src/consensus_state.rs

```diff
@@ -70,6 +73,91 @@ pub struct ConsensusState {
 }
 
 impl ConsensusState {
+    // TODO: read from config
+    pub fn new() -> Self {
```

Comment:
better to impl Default trait when it has hard coded values

Reply:
Can't use in this case will raise conflict
![Screenshot (387)](https://user-images.githubusercontent.com/8589224/148321331-f6e71746-feea-490a-aeb5-a4ca94a3e66c.png)

Reply:
Then u can remove the derived Default impl

Reply:
If remove derived Default will affect this method [line](https://github.com/ChainSafe/mina-rs/blob/6935fc291c3551498f12d1758f407bbef79068cb/base/src/consensus_state.rs#L75
![Screenshot (394)](https://user-images.githubusercontent.com/8589224/148472817-6669857a-ff19-4f47-9bef-003035ba5fa2.png)
)

Reply:
If take this method out will cause test_protocol_state_chain_top to fail.
![Screenshot (395)](https://user-images.githubusercontent.com/8589224/148473643-591a7b7c-c532-43ae-ad6d-7d0f6ef84776.png)

Reply:
Sry but I still don't understand the root cause

Reply:
Affect this [line](https://github.com/ChainSafe/mina-rs/blob/68acc0c4021d32de6f2f3af11cdc7927b2bfd975/consensus/src/common.rs#L50)

Reply:
Sry can u plz elaborate on why this is not possible? I think 'affect this line' is the result instead of the root cause,

Reply:
The left blockchain_length: Length(0) compare right  blockchain_length: Length(1) if we use default implementation.
```
 left: `Some(ProtocolState { previous_state_hash: StateHash(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), body: ProtocolStateBody { genesis_state_hash: StateHash(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), blockchain_state: BlockchainState { staged_ledger_hash: StagedLedgerHash { non_snark: NonSnarkStagedLedgerHash { ledger_hash: LedgerHash(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), aux_hash: AuxHash([]), pending_coinbase_aux: PendingCoinbaseAuxHash([]) }, pending_coinbase_hash: CoinBaseHash(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])) }, snarked_ledger_hash: SnarkedLedgerHash(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), genesis_ledger_hash: SnarkedLedgerHash(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), snarked_next_available_token: TokenId(0), timestamp: BlockTime(0) }, consensus_state: ConsensusState { blockchain_length: Length(0), epoch_count: Length(0), min_window_density: Length(77), sub_window_densities: [Length(1), Length(7), Length(7), Length(7), Length(7), Length(7), Length(7), Length(7), Length(7), Length(7), Length(7)], last_vrf_output: VrfOutputTruncated([13]), total_currency: Amount(805385692840039233), curr_global_slot: GlobalSlot { slot_number: GlobalSlotNumber(0), slots_per_epoch: Length(7140) }, global_slot_since_genesis: GlobalSlotNumber(0), staking_epoch_data: EpochData { ledger: EpochLedger { hash: LedgerHash(BaseHash([133, 246, 174, 16, 42, 48, 61, 19, 82, 234, 190, 184, 70, 31, 98, 164, 122, 104, 123, 109, 205, 231, 98, 51, 185, 32, 113, 104, 160, 172, 247, 44])), total_currency: Amount(805385692840039233) }, seed: EpochSeed(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), start_checkpoint: StateHash(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), lock_checkpoint: StateHash(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), epoch_length: Length(1) }, next_epoch_data: EpochData { ledger: EpochLedger { hash: LedgerHash(BaseHash([133, 246, 174, 16, 42, 48, 61, 19, 82, 234, 190, 184, 70, 31, 98, 164, 122, 104, 123, 109, 205, 231, 98, 51, 185, 32, 113, 104, 160, 172, 247, 44])), total_currency: Amount(805385692840039233) }, seed: EpochSeed(BaseHash([37, 127, 117, 208, 167, 148, 99, 11, 218, 242, 217, 7, 37, 244, 201, 93, 161, 26, 189, 134, 47, 184, 162, 248, 138, 185, 76, 65, 250, 11, 141, 53])), start_checkpoint: StateHash(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), lock_checkpoint: StateHash(BaseHash([232, 149, 55, 214, 162, 25, 166, 75, 189, 7, 127, 231, 219, 235, 171, 161, 183, 9, 227, 123, 211, 166, 117, 103, 24, 130, 127, 33, 8, 154, 124, 2])), epoch_length: Length(2) }, has_ancestor_in_same_checkpoint_window: true, block_stake_winner: PublicKey { poly: CompressedCurvePoint { x: [15, 72, 198, 91, 210, 95, 133, 243, 228, 234, 78, 254, 190, 183, 91, 121, 123, 215, 67, 96, 59, 224, 75, 78, 173, 132, 86, 152, 183, 107, 211, 49], is_odd: false } }, block_creator: PublicKey { poly: CompressedCurvePoint { x: [15, 72, 198, 91, 210, 95, 133, 243, 228, 234, 78, 254, 190, 183, 91, 121, 123, 215, 67, 96, 59, 224, 75, 78, 173, 132, 86, 152, 183, 107, 211, 49], is_odd: false } }, coinbase_receiver: PublicKey { poly: CompressedCurvePoint { x: [15, 72, 198, 91, 210, 95, 133, 243, 228, 234, 78, 254, 190, 183, 91, 121, 123, 215, 67, 96, 59, 224, 75, 78, 173, 132, 86, 152, 183, 107, 211, 49], is_odd: false } }, supercharge_coinbase: true }, constants: ProtocolConstants { k: Length(0), slots_per_epoch: Length(0), slots_per_sub_window: Length(0), delta: Delta(0), genesis_state_timestamp: BlockTime(0) } } })`,
 right: `Some(ProtocolState { previous_state_hash: StateHash(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), body: ProtocolStateBody { genesis_state_hash: StateHash(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), blockchain_state: BlockchainState { staged_ledger_hash: StagedLedgerHash { non_snark: NonSnarkStagedLedgerHash { ledger_hash: LedgerHash(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), aux_hash: AuxHash([]), pending_coinbase_aux: PendingCoinbaseAuxHash([]) }, pending_coinbase_hash: CoinBaseHash(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])) }, snarked_ledger_hash: SnarkedLedgerHash(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), genesis_ledger_hash: SnarkedLedgerHash(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), snarked_next_available_token: TokenId(0), timestamp: BlockTime(0) }, consensus_state: ConsensusState { blockchain_length: Length(1), epoch_count: Length(0), min_window_density: Length(77), sub_window_densities: [Length(1), Length(7), Length(7), Length(7), Length(7), Length(7), Length(7), Length(7), Length(7), Length(7), Length(7)], last_vrf_output: VrfOutputTruncated([13]), total_currency: Amount(805385692840039233), curr_global_slot: GlobalSlot { slot_number: GlobalSlotNumber(0), slots_per_epoch: Length(7140) }, global_slot_since_genesis: GlobalSlotNumber(0), staking_epoch_data: EpochData { ledger: EpochLedger { hash: LedgerHash(BaseHash([133, 246, 174, 16, 42, 48, 61, 19, 82, 234, 190, 184, 70, 31, 98, 164, 122, 104, 123, 109, 205, 231, 98, 51, 185, 32, 113, 104, 160, 172, 247, 44])), total_currency: Amount(805385692840039233) }, seed: EpochSeed(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), start_checkpoint: StateHash(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), lock_checkpoint: StateHash(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), epoch_length: Length(1) }, next_epoch_data: EpochData { ledger: EpochLedger { hash: LedgerHash(BaseHash([133, 246, 174, 16, 42, 48, 61, 19, 82, 234, 190, 184, 70, 31, 98, 164, 122, 104, 123, 109, 205, 231, 98, 51, 185, 32, 113, 104, 160, 172, 247, 44])), total_currency: Amount(805385692840039233) }, seed: EpochSeed(BaseHash([37, 127, 117, 208, 167, 148, 99, 11, 218, 242, 217, 7, 37, 244, 201, 93, 161, 26, 189, 134, 47, 184, 162, 248, 138, 185, 76, 65, 250, 11, 141, 53])), start_checkpoint: StateHash(BaseHash([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), lock_checkpoint: StateHash(BaseHash([232, 149, 55, 214, 162, 25, 166, 75, 189, 7, 127, 231, 219, 235, 171, 161, 183, 9, 227, 123, 211, 166, 117, 103, 24, 130, 127, 33, 8, 154, 124, 2])), epoch_length: Length(2) }, has_ancestor_in_same_checkpoint_window: true, block_stake_winner: PublicKey { poly: CompressedCurvePoint { x: [15, 72, 198, 91, 210, 95, 133, 243, 228, 234, 78, 254, 190, 183, 91, 121, 123, 215, 67, 96, 59, 224, 75, 78, 173, 132, 86, 152, 183, 107, 211, 49], is_odd: false } }, block_creator: PublicKey { poly: CompressedCurvePoint { x: [15, 72, 198, 91, 210, 95, 133, 243, 228, 234, 78, 254, 190, 183, 91, 121, 123, 215, 67, 96, 59, 224, 75, 78, 173, 132, 86, 152, 183, 107, 211, 49], is_odd: false } }, coinbase_receiver: PublicKey { poly: CompressedCurvePoint { x: [15, 72, 198, 91, 210, 95, 133, 243, 228, 234, 78, 254, 190, 183, 91, 121, 123, 215, 67, 96, 59, 224, 75, 78, 173, 132, 86, 152, 183, 107, 211, 49], is_odd: false } }, supercharge_coinbase: true }, constants: ProtocolConstants { k: Length(0), slots_per_epoch: Length(0), slots_per_sub_window: Length(0), delta: Delta(0), genesis_state_timestamp: BlockTime(0) } } })`
```

Reply:
At [line](https://github.com/ChainSafe/mina-rs/blob/6935fc291c3551498f12d1758f407bbef79068cb/consensus/tests/common.rs#L50)
The one below line 50 passes the test
![Screenshot (402)](https://user-images.githubusercontent.com/8589224/148904065-e57123b4-f507-4f32-91ec-1308e60a09a9.png)

Reply:
Fixed.